Add note about transport mTLS

DaveCTurner opened the pull request:
Until recently, public CAs would issue certificates with an Extended Key
Usage set that includes the `clientAuth` value, allowing these
certificates to be used for mTLS. Nonetheless it is a mistake to use
such certificates for mTLS. To prevent users from continuing to make
this mistake, all certificates issued by public CAs will soon omit the
`clientAuth` usage value.

Elasticsearch will by default use mTLS for inter-node connections, and
we've recently encountered some users who have been obtaining their
transport certificates from such public CAs and mistakenly using them
for mTLS. This commit adds some documentation clarifying the security
model and giving clearer recommendations in this area.

DaveCTurner committed:
Add note about transport mTLS

Until recently, public CAs would issue certificates with an Extended Key
Usage set that includes the `clientAuth` value, allowing these
certificates to be used for mTLS. Nonetheless it is a mistake to use
such certificates for mTLS. To prevent users from continuing to make
this mistake, all certificates issued by public CAs will soon omit the
`clientAuth` usage value.

Elasticsearch will by default use mTLS for inter-node connections, and
we've recently encountered some users who have been obtaining their
transport certificates from such public CAs and mistakenly using them
for mTLS. This commit adds some documentation clarifying the security
model and giving clearer recommendations in this area.

DaveCTurner committed:
Move wall of text to section top

DaveCTurner committed:
Wordsmithery

DaveCTurner committed:
tweak

DaveCTurner committed:
Merge branch 'main' into 2025/11/06/transport-mtls-docs

DaveCTurner committed:
Fix incorrect negation

DaveCTurner committed:
Merge branch 'main' into 2025/11/06/transport-mtls-docs

DaveCTurner committed:
Add some callouts

DaveCTurner committed:
More security words

DaveCTurner commented on docs/reference/elasticsearch/configuration-reference/security-settings.md:
```diff
@@ -1911,7 +1911,15 @@ PKCS#12 files are configured in the same way as Java keystore files:
+
+The security requirements for transport certificates (as defined by the `xpack.security.transport.ssl.*` settings) are significantly different from the security requirements for HTTP certificates (as defined by the `xpack.security.http.ssl.*` settings). HTTP connections do not generally use mTLS since HTTP has its own authentication mechanisms, so HTTP certificates do not usually need to include  ...
+
+If your environment has some other way to prevent unauthorized node-to-node connections, you may prefer not to use mTLS for transport connections. In this case, turn mTLS off in this context by setting `xpack.security.transport.ssl.client_authentication: none`. You may still use (non-mutual) TLS to ensure the confidentiality and integrity of node-to-node traffic by setting `xpack.security.transpo ...
```
++ makes sense, I made it more opinionated

DaveCTurner committed:
Optional is also bad

DaveCTurner committed:
Iter

DaveCTurner committed:
Iter

DaveCTurner committed:
Call out common mistake

DaveCTurner committed:
Remove suggestion to use public CAs if mTLS disabled

DaveCTurner committed:
Context is overloaded

shainaraskas reviewed (commented):
this is a lot of dense best practice stuff. consider moving most of this guidance info into the instructional docs, and just hint here that there are some security best practices to follow (and link out) 

possible locations (didn't look super deeply, but I can if desired)

https://www.elastic.co/docs/deploy-manage/security/secure-cluster-communications
https://www.elastic.co/docs/deploy-manage/security/set-up-basic-security
https://www.elastic.co/docs/deploy-manage/security/self-tls (or a new child page here)

DaveCTurner committed:
Merge branch 'main' into 2025/11/06/transport-mtls-docs

shainaraskas reviewed (dismissed):
wont' block this from being merged, but I still think there's room in our other docs for this info even if our current content doesn't have E2E procedures. my biggest concern here is visibility/discoverability because this information is urgent / causing problems. 

I'd consider creating an H2 above the `generate` step on this page: https://www.elastic.co/docs/deploy-manage/security/set-up-basic-security titled something like `Considerations for using an external CA` (I believe this page is the entrypoint for setting up external certs).

if you would prefer to ship this here, we should consider adding a warning on that page (and maybe some of the other TLS encryption pages) that links to this, so people can more easily discover it. let me know what you prefer.

wonder if this info is also relevant to [ECK transport certs](https://www.elastic.co/docs/deploy-manage/security/k8s-transport-settings).

tvernum reviewed (dismissed):
Thanks!

shainaraskas commented:
@DaveCTurner working on this now

shainaraskas commented on docs/reference/elasticsearch/configuration-reference/security-settings.md:
```diff
@@ -1911,7 +1911,27 @@ PKCS#12 files are configured in the same way as Java keystore files:
+
+::::{warning}
+If you turn off mTLS by setting `xpack.security.transport.ssl.client_authentication` to `optional` or `none` then anyone with network access to a node in your {{es}} cluster will be able to establish a transport connection with that node. A malicious actor with this access may be able to use such a transport connection to invoke certain system-internal APIs, some of which may allow them to read o ...
+::::
```
@DaveCTurner I've [created a PR](https://github.com/elastic/docs-content/pull/3932) to move the bulk of the recommendations to the instructional docs, and also add lots of warnings in any spot where people might be assessing using their own CA. as a result, I think we can whittle this down to the following. the linked PR will need to ship first but this can follow it shortly after.

```suggestion
By default, {{es}} uses mutual TLS (mTLS) to secure node-to-node transport connections within a cluster. With mTLS, data is encrypted in transit and both nodes must present valid certificates when connecting. Each node requires that certificates be issued by a trusted certificate authority, ensuring that only authorized nodes can connect. [Learn about configuring node-to-node mTLS](docs-content://deploy-manage/security/secure-cluster-communications.md#encrypt-internode-communication).

:::{warning}
Transport connections between {{es}} nodes are security-critical and you must protect them carefully. Malicious actors who can observe or interfere with node-to-node transport traffic can read or modify cluster data. A malicious actor who can establish a transport connection might be able to invoke system-internal APIs, including APIs that read or modify cluster data.

If you choose to issue node transport certificates using third-party tools, then carefully review [Considerations for using an external CA for transport layer security](docs-content://deploy-manage/security/self-tls-considerations.md) to ensure that the certificates that you provide meet the security requirements for transport connections.

If your environment has some other way to prevent unauthorized node-to-node connections, you might prefer not to use mTLS for transport connections. [Learn more](docs-content://deploy-manage/security/self-tls-considerations.md#turn-off-mtls).
:::
```

eedugon reviewed (commented):
Great improvements in this PR, but in one area I think it adds the wrong impression that for mTLS (in general) the server certs need both `clientAuth` and `serverAuth`, when that only applies to the specifics of Elasticsearch transport, where the same cert is used by the node acting as client and a server.

I've suggested a few small changes that should clarify this a bit better (hopefully) without changing the essence.

The proposals are for:
- Ensure we specify `{{es}} transport` when we say something that could be felt like `a server cert without clientAuth extension is not suitable for mTLS`.
- Explain why Elasticsearch transport cert requires both extension values or omit the extension (added a `because` short clause).
- Rephrase a bit the HTTP related paragrapth to remove the relation with mTLS (public CA-signed certs could be used for Elasticsearch HTTP server even if mTLS is required by the user).

eedugon commented:
> If it turns out that more clarity is needed in this area then we can refine it later, but I think in that case it'll need more substantial rework than replacing "mTLS" with "Elasticsearch transport mTLS" as you've suggested.

Well, the suggestion is because implying that a server cert without `clientAuth` is not valid for mTLS could be misleading, where the fact is that it's not valid for us, for the Elasticsearch transport, and because the entity is a client and a server at the same time.

I'm trying to 1st help, and 2nd trying to provide the smallest changes possible to make the content accurate. But anyway, whatever you feel it's better. We could refine at a later stage, for sure.

DaveCTurner committed:
Merge branch 'main' into 2025/11/06/transport-mtls-docs

DaveCTurner committed:
Rework

https://github.com/elastic/docs-content/pull/3932 moved most of this to
`docs-content`

DaveCTurner committed:
Merge branch 'main' into 2025/11/06/transport-mtls-docs

shainaraskas reviewed (approved):
lgtm 🎉
